### PR TITLE
feat(x/data)!: enable off-chain coordination of supported algorithms

### PR DESCRIPTION
**Pull request by aaronc**
## Description

This PR introduces `regen.data.v2` which:
1. does not validate digest, canonicalization or merkle tree algorithms
2. does not use an enum for raw data extensions

This allows off-chain coordination of these values so that extensions can be made without requiring an upgrade.

`regen/data/v2/types.proto` SHOULD still be used as a registry of agreed upon algorithms, but this can be done between releases without requiring chain upgrades.

This PR deprecates the `regen.data.v1` proto files and bumps the x/data go module to v3.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: scripts/protocgen.sh [hunk 1]
```diff
@@ -30,6 +30,4 @@ cd ..
 cp -r github.com/regen-network/regen-ledger/* ./
 rm -rf github.com
 
-go mod tidy
```
Comment (aaronc) [inline scripts/protocgen.sh:33]: This was breaking `make proto-gen` and can be done manually by the user

File: scripts/protocgen2.sh [hunk 2]
```diff
@@ -5,7 +5,7 @@ set -eo pipefail
 protoc_gen_install() {
   go install github.com/cosmos/cosmos-proto/cmd/protoc-gen-go-pulsar@latest #2>/dev/null
   go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@latest #2>/dev/null
-  go install github.com/cosmos/cosmos-sdk/orm/cmd/protoc-gen-go-cosmos-orm@latest #2>/dev/null
+  go install github.com/cosmos/cosmos-sdk/orm/cmd/protoc-gen-go-cosmos-orm@v1.0.0-alpha.12 #2>/dev/null
```
Comment (aaronc) [inline scripts/protocgen2.sh:8]: This was breaking `make proto-gen`

File: x/data/iri_test.go [hunk 3]
```diff
@@ -445,3 +638,12 @@ func TestParseIRI(t *testing.T) {
 		})
 	}
 }
+
+func TestName(t *testing.T) {
```
Comment (blushi) [inline x/data/iri_test.go:642]: to remove?

**Comment by aaronc** [comment]
> Looks good, just a few nits and pending lint errors
> 
> Should I just use version 2 moving forward in regen-server/iri-gen, in particular in my PR: https://github.com/regen-network/regen-server/pull/423

We have no idea how long this will take to go into production so for now we need to stick with v1.

**Comment by clevinson** [comment]
Thinking about the long-term effects of a versioning name change like this, I'm wondering if we'd be better off with one of the following:
1. using the pre-existing versioning system given by the protocol paths (e.g. just bumping this to `proto/regen/data/v2..`)
2. Updating the existing implementation of the current `Graph` and `Raw` types (in `promo/regen/data/v1...`) to match the desired implementation (what this PR current does in `GraphV2` and `RawV2`).

Conceptually I'm in favor of this general approach (not validating enums on-chain to enable for faster developing & off-chain coordination).  But given that we have such little usage of the data module to date, I think we may be able to skirt around what would otherwise be the strictly correct process for API upgrades, in favor of something that's a bit cleaner for future users.

**Comment by aaronc** [comment]
> Thinking about the long-term effects of a versioning name change like this, I'm wondering if we'd be better off with one of the following:
> 
> 1. using the pre-existing versioning system given by the protocol paths (e.g. just bumping this to `proto/regen/data/v2..`)
> 2. Updating the existing implementation of the current `Graph` and `Raw` types (in `promo/regen/data/v1...`) to match the desired implementation (what this PR current does in `GraphV2` and `RawV2`).

I would be open to either approach. I've checked data module events on mainnet and so far there is no usage. So until we start using it (which is soon) breakage isn't really a concern. With approach 1) we would simply stop supporting v1 I think to keep things simple. Approach 2) is maybe okay since there are no other users but it is a bit un-kosher. Don't have any really strong opinions between the two though. @blushi any thoughts?

**Comment by blushi** [comment]
> > Thinking about the long-term effects of a versioning name change like this, I'm wondering if we'd be better off with one of the following:
> > 
> > 1. using the pre-existing versioning system given by the protocol paths (e.g. just bumping this to `proto/regen/data/v2..`)
> > 2. Updating the existing implementation of the current `Graph` and `Raw` types (in `promo/regen/data/v1...`) to match the desired implementation (what this PR current does in `GraphV2` and `RawV2`).
> 
> I would be open to either approach - and I think this would greatly simply the implementation. I've checked data module events on mainnet and so far there is no usage. So until we start using it (which is soon) breakage isn't really a concern. With approach 1) we would simply stop supporting v1 I think to keep things simple. Approach 2) is maybe okay since there are no other users but it is a bit un-kosher. Don't have any really strong opinions between the two though. @blushi any thoughts?

I think I'd rather go with approach 1) because v1 hasn't been explicitly used on mainnet but it has been in our implementation within the regen server for generating IRIs for credit classes, projects and batches metadata so that would feel cleaner to have a proper version upgrade.

**Comment by aaronc** [comment]
Okay. I'll update this PR to change the proto version to v2 and drop support for v1, unless there are any objections

**Comment by aaronc** [comment]
@clevinson @blushi I've updated this to v2 (and the go.mod to v3) and it's R4R again. The diff is much larger because of breaking changes, but the code changes are actually somewhat simpler.

File: x/data/iri.go [hunk 4]
```diff
@@ -43,18 +45,15 @@ func (chr ContentHash_Raw) ToIRI() (string, error) {
 	bz[1] = byte(chr.DigestAlgorithm)
 	copy(bz[2:], chr.Hash)
 	hashStr := base58.CheckEncode(bz, iriVersion0)
```
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/iri.go:47]: I think it's actually okay to keep the IRI version at 0. Nothing substantively changed in how the IRIs are constructed.

**Comment by aaronc** [comment]
@blushi I've addressed all your comments, except that I'd prefer to keep the IRI on version 0 and updated the comments to reflect that.

File: x/data/iri.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (blushi) [reply x/data/iri.go:45]: right, seems fine now that we've bumped to v2

File: x/data/types.go [hunk 5]
```diff
@@ -26,27 +26,43 @@ func (ch ContentHash) Validate() error {
...
+	}
+
+	if extLen > 6 {
+		return sdkerrors.ErrInvalidRequest.Wrapf("file extension cannot be longer than 6 characters")
+	}
+
+	// check that ext is all lowercase or numeric
+	for _, c := range ext {
+		if c < '0' || c > '9' && c < 'a' || c > 'z' {
+			return sdkerrors.ErrInvalidRequest.Wrapf("file extension must be all lowercase or numeric")
```
Comment (blushi) [inline x/data/types.go:48]: can we add a test case for this in iri_test.go?

**Review by blushi — APPROVED** [review]
LGTM, pre-approving, pending one test case to add

File: x/data/types.go (hunk 5 above)
Comment (aaronc) [reply x/data/types.go:48]: Added a case

**Comment by aaronc** [comment]
@clevinson could you take another quick look at this?

**Review by clevinson — APPROVED** [review]
Sorry this took so long. lgtm!

**Pull request merged**